Python bindings for EclIO classes

tskille opened the pull request:
This PR include python bindings for Eclipse IO classes EclFile, ERst, EGrid, ERft, ESmry and EclOutput

akva2 commented on python/cxx/eclipse_io.cpp:
```diff
@@ -13,26 +13,75 @@ namespace py = pybind11;
+    for (auto element : array_list)
+        arrayNameList.push_back(std::get<0>(element));
+
+    std::vector<std::string>::iterator it = arrayNameList.begin();
```
let's avoid all the copies.

```
    size_t cidx = 0;
    auto it = std::find_if(array_list.begin(), array_list.end(),
                           [&cidx, &array_name, occurence](const EclEntry& entry)
                           {
                              if (std::get<0>(entry) == array_name)
                                  ++cidx;

                              return cidx == occurence + 1;
                           });

    return std::distance(array_list.begin(), it);
```

tskille replied:
done

akva2 commented on opm/io/eclipse/ERft.hpp:
```diff
@@ -63,18 +67,24 @@ class ERft : public EclFile
     RftReportList rftReportList;
 
-    std::map<std::pair<std::string,RftDate>,int> reportIndex;  //  mapping report index to wellName and date (tupe)
+    std::map<std::tuple<std::string,RftDate>,int> reportIndex;  //  mapping report index to wellName and date (tupe)
```
this member shadows with every single method taking reportIndex as a parameter. rename it, e.g. 'reportIndices', or change the parameter names in the methods.

tskille replied:
done

akva2 commented on opm/io/eclipse/ERst.hpp:
```diff
@@ -51,6 +51,7 @@ class ERst : public EclFile
     }
 
     int count(const std::string& name, int reportStepNumber) const; 
+    int numberOfReportSteps() const { return seqnum.size(); };
```
seqnum is  a vector, it's size is size_t, not int. i do not see see why we would want make it signed and truncated.

bska replied:
> it's size is size_t,

It actually isn't.  The return value from `vector<T>::size()` is really
```C++
std::vector<T>::size_type
```
which **happens to be** `std::size_t` in many implementations.

tskille replied:
done

akva2 commented on opm/io/eclipse/ESmry.hpp:
```diff
@@ -41,10 +41,14 @@ class ESmry
 
     int timestepIdxAtReportstepStart(const int reportStep) const;
 
+    int numberOfTimeSteps() const { return param[0].size(); }
```
why int?

tskille replied:
done

akva2 commented on python/cxx/eclipse_io.cpp:
```diff
@@ -13,29 +20,268 @@ namespace py = pybind11;
 
-py::array get_vector(Opm::EclIO::EclFile * file_ptr, std::size_t array_index) {
+using npArray = std::tuple<py::array, Opm::EclIO::eclArrType>;
+using EclEntry = std::tuple<std::string, Opm::EclIO::eclArrType, long int>;
```
any reason why this is a long int? it's icky for portability.

tskille replied:
As part of supporting huge arrays , PR #1587 "Adding support for larger EclFile arrays ", EclEntry was changed in EclFile.hpp from int to long int

```C++
using EclEntry = std::tuple<std::string, eclArrType, long int>;
```

akva2 replied:
well as discussed in email windows has a surprise for you. it's better to use int64_t.

tskille replied:
I'm planning to go through the EclIO classes and replace long int with int64_t and from unsigned long int with uint64_t.  

I will preferre to do this in a separate PR. I will start with this when this is completed.

tskille commented:
Thanks @akva2. I have now addressed all comments you had and hopefully this is now ready to be merge. When This is merged, I'll start working on replacing **long int** with **int64_t** and **unsigned long int** with **uint64_t**.

akva2 commented:
jenkins build this please

akva2 commented:
quell https://ci.opm-project.org/job/opm-common-PR-builder/1644/gcc/new/

akva2 commented:
jenkins build this please

akva2 commented:
https://ci.opm-project.org/job/opm-common-PR-builder/1669/gcc/new/source.0e9d384b-6277-4427-b662-7ad9d15ec884/#125

akva2 commented:
jenkins build this please
